hazelcast/src/main/java/com/hazelcast/map/impl/record/CachedDataRecordWithStats.java

```diff
@@ -18,10 +18,14 @@
 
 import com.hazelcast.nio.serialization.Data;
 
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+
 /**
  * CachedDataRecordWithStats.
  */
 class CachedDataRecordWithStats extends DataRecordWithStats {
+    private static final AtomicReferenceFieldUpdater<CachedDataRecordWithStats, Object> CACHED_VALUE_UPDATER =
+            AtomicReferenceFieldUpdater.newUpdater(CachedDataRecordWithStats.class, Object.class, "cachedValue");
 
     private transient volatile Object cachedValue;
```

Comment:
It would be a good thing to add documentation here explaining that the cachedValue not only contains the value or null, but could also have a thread/thread-wrapper to indicate it is currently being locked.

If somebody would just read out the cachedValue, then bogus values could appear out of thin air.